random-fu/src/Data/Random/Sample.hs

```diff
@@ -1,46 +1,50 @@
 {-# LANGUAGE
         MultiParamTypeClasses,
-        FlexibleInstances, FlexibleContexts, 
+        FlexibleInstances, FlexibleContexts,
         IncoherentInstances
   #-}
 
 {-# OPTIONS_GHC -fno-warn-simplifiable-class-constraints #-}
 
 module Data.Random.Sample where
 
-import Control.Monad.State 
+import Control.Monad.State
+import Control.Monad.Reader
 import Data.Random.Distribution
 import Data.Random.Lift
 import Data.Random.RVar
-import Data.Random.Source
-import Data.Random.Source.Std
+
+import System.Random.Stateful
 
 -- |A typeclass allowing 'Distribution's and 'RVar's to be sampled.  Both may
 -- also be sampled via 'runRVar' or 'runRVarT', but I find it psychologically
 -- pleasing to be able to sample both using this function, as they are two
 -- separate abstractions for one base concept: a random variable.
 class Sampleable d m t where
     -- |Directly sample from a distribution or random variable, using the given source of entropy.
-    sampleFrom :: RandomSource m s => s -> d t -> m t
+    sampleFrom :: StatefulGen g m => g -> d t -> m t
 
 instance Distribution d t => Sampleable d m t where
-    sampleFrom src d = runRVarT (rvar d) src
+    sampleFrom gen d = runRVarT (rvar d) gen
 
 -- This instance overlaps with the other, but because RVarT is not a Distribution there is no conflict.
 instance Lift m n => Sampleable (RVarT m) n t where
-    sampleFrom src x = runRVarT x src
+    sampleFrom gen x = runRVarT x gen
 
 -- |Sample a random variable using the default source of entropy for the
 -- monad in which the sampling occurs.
-sample :: (Sampleable d m t, MonadRandom m) => d t -> m t
-sample = sampleFrom StdRandom
+sample :: (Sampleable d m t, StatefulGen g m, MonadReader g m) => d t -> m t
```

Comment:
I think I already mentioned that I don't know of an instance here but maybe I am not thinking hard enough.

Reply:
Same answer applies here:

```haskell
sample :: PrimMonad m => Gamma Float -> ReaderT (Gen (PrimState m)) m Float
```

This is exactly what `MonadRandom` allowed us to do. 

Another example is this:

```haskell
sample :: Uniform Int -> ReaderT (StageGenM StdGen) (State StdGen) Int
```